share/getters/cascade.go

```diff
@@ -0,0 +1,151 @@
+package getters
+
+import (
+	"context"
+	"time"
+
+	"go.uber.org/multierr"
+
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/nmt/namespace"
+	"github.com/celestiaorg/rsmt2d"
+)
+
+// CascadeGetter implements custom share.Getter that composes multiple Getter implementations in
+// "cascading" order.
+//
+// See cascade func for details on cascading.
+type CascadeGetter struct {
+	interval time.Duration
+
+	getters []share.Getter
+}
+
+// NewCascadeGetter instantiates a new CascadeGetter from given share.Getters with given interval.
+func NewCascadeGetter(getters []share.Getter, interval time.Duration) *CascadeGetter {
+	return &CascadeGetter{
+		interval: interval,
+		getters:  getters,
+	}
+}
+
+// GetShare gets a share from any of registered share.Getters in cascading order.
+func (cg *CascadeGetter) GetShare(ctx context.Context, root *share.Root, row, col int) (share.Share, error) {
+	get := func(ctx context.Context, get share.Getter) (share.Share, error) {
+		return get.GetShare(ctx, root, row, col)
+	}
+	return cascadeGetters(ctx, cg.getters, cg.interval, get)
+}
+
+// GetEDS gets a full EDS from any of registered share.Getters in cascading order.
+func (cg *CascadeGetter) GetEDS(ctx context.Context, root *share.Root) (*rsmt2d.ExtendedDataSquare, error) {
+	get := func(ctx context.Context, get share.Getter) (*rsmt2d.ExtendedDataSquare, error) {
+		return get.GetEDS(ctx, root)
+	}
+	return cascadeGetters(ctx, cg.getters, cg.interval, get)
+}
+
+// GetSharesByNamespace gets NamespacedShares from any of registered share.Getters in cascading
+// order.
+func (cg *CascadeGetter) GetSharesByNamespace(
+	ctx context.Context,
+	root *share.Root,
+	id namespace.ID,
+) (share.NamespacedShares, error) {
+	get := func(ctx context.Context, get share.Getter) (share.NamespacedShares, error) {
+		return get.GetSharesByNamespace(ctx, root, id)
+	}
+	return cascadeGetters(ctx, cg.getters, cg.interval, get)
+}
+
+func cascadeGetters[V any](
+	ctx context.Context,
+	getters []share.Getter,
+	interval time.Duration,
+	get func(context.Context, share.Getter) (V, error),
+) (V, error) {
+	fns := make([]func(context.Context) (V, error), 0, len(getters))
+	for _, getter := range getters {
+		getter := getter // required for the same reason we do this when we launch goroutines in the loop
+		fns = append(fns, func(ctx context.Context) (V, error) {
+			return get(ctx, getter)
+		})
+	}
+
+	return cascade[V](ctx, fns, interval)
+}
+
+// cascade implements a cascading retry algorithm for getting a value from multiple sources
+// sorted by priority. Cascading implies trying the sources one-by-one in the given order with the
+// given interval until either:
```

Comment:
Can you please add that the given timeout interval does not cancel the last source, but starts the next one and results can still come from the previous source?

Reply:
Actually, this is implied by `All of the sources errors`,does not hurt to clarify that though